test/ret_web/controllers/api/credentials_controller_test.exs

```diff
@@ -0,0 +1,362 @@
+
+    assert conn
+           |> put_auth_header_for_account(account)
+           |> patch(credentials_path(conn, :update, Map.get(creds, "id"), revoke: true))
+           |> json_response(200)
+           |> Map.get("credentials")
+           |> hd()
+           |> Map.get("is_revoked")
+  end
+
+  test "Non-admins cannot revoke credentials they do not own", %{
+    account: account,
+    conn: conn,
+    create: create
+  } do
+    another_account = create_account("another_account", false)
+
+    creds =
+      conn
+      |> put_auth_header_for_account(another_account)
+      |> post(create, params(valid_scopes(), :account))
+      |> json_response(200)
+      |> Map.get("credentials")
+      |> hd()
+
+    refute Map.get(creds, "is_revoked")
+
+    sid = Map.get(creds, "id")
+
+    assert conn
+           |> put_auth_header_for_account(account)
+           |> patch(credentials_path(conn, :update, sid, revoke: true))
+           |> json_response(401)
+
+    refute conn
+           |> put_auth_header_for_account(another_account)
+           |> get(credentials_path(conn, :show, sid))
+           |> json_response(200)
+           |> Map.get("credentials")
+           |> hd()
+           |> Map.get("is_revoked")
+
+    assert conn
+           |> put_auth_header_for_account(another_account)
+           |> patch(credentials_path(conn, :update, sid, revoke: true))
+           |> json_response(200)
+           |> Map.get("credentials")
+           |> hd()
+           |> Map.get("is_revoked")
+  end
+
+  test "Admins can revoke any credentials", %{
+    admin_account: admin_account,
+    account: account,
+    conn: conn,
+    create: create
+  } do
+    another_admin_account = create_account("another_admin", true)
+
+    # Admins can revoke app tokens they do not own
+    creds =
+      conn
+      |> put_auth_header_for_account(another_admin_account)
+      |> post(create, params(valid_scopes(), :app))
+      |> json_response(200)
+      |> Map.get("credentials")
+      |> hd()
+
+    refute Map.get(creds, "is_revoked")
+
+    assert conn
+           |> put_auth_header_for_account(admin_account)
+           |> patch(credentials_path(conn, :update, Map.get(creds, "id"), revoke: true))
+           |> json_response(200)
+           |> Map.get("credentials")
+           |> hd()
+           |> Map.get("is_revoked")
+
+    # Admins can revoke account tokens they do not own
+    creds =
+      conn
+      |> put_auth_header_for_account(account)
+      |> post(create, params(valid_scopes(), :account))
+      |> json_response(200)
+      |> Map.get("credentials")
+      |> hd()
+
+    refute Map.get(creds, "is_revoked")
+
+    assert conn
+           |> put_auth_header_for_account(admin_account)
+           |> patch(credentials_path(conn, :update, Map.get(creds, "id"), revoke: true))
+           |> json_response(200)
+           |> Map.get("credentials")
+           |> hd()
+           |> Map.get("is_revoked")
+  end
+end
```

Comment:
:+1: nice tests!

Reply:
Thanks :)

I have not wrapped my head around or studied up on best practices for testing elixir + phoenix apps. Here I test an interface (`CredentialsController`) and the model (`Credentials`) together, which I could have done this separately. I am happy enough with it as it is, but want to be better about this in the future. This seems like it'll be increasingly important as we have alternative interfaces, transports, and views (REST over HTTP, graphql, websockets passing json).